indra/newview/llagentbenefits.cpp

```diff
@@ -142,9 +161,54 @@ S32 LLAgentBenefits::getTextureUploadCost() const
 	return m_texture_upload_cost;
 }
 
-S32 LLAgentBenefits::get2KTextureUploadCost() const
+S32 LLAgentBenefits::getTextureUploadCost(const LLViewerTexture* tex) const
+{
+    if (tex)
+    {
+        S32 area = tex->getFullHeight() * tex->getFullWidth();
+        if (area >= MIN_2K_TEXTURE_AREA)
+        {
+            return get2KTextureUploadCost(area);
+        }
+        else
+        {
+            return getTextureUploadCost();
+        }
+    }
+    return getTextureUploadCost();
+}
+
+S32 LLAgentBenefits::getTextureUploadCost(const LLImageBase* tex) const
+{
+    if (tex)
+    {
+        S32 area = tex->getHeight() * tex->getWidth();
+        if (area >= MIN_2K_TEXTURE_AREA)
+        {
+            return get2KTextureUploadCost(area);
+        }
+        else
+        {
+            return getTextureUploadCost();
+        }
+    }
+    return getTextureUploadCost();
+}
+
+S32 LLAgentBenefits::get2KTextureUploadCost(S32 area) const
 {
-    return m_2k_texture_upload_cost;
+    if (m_2k_texture_upload_cost.empty())
+    {
+        return m_texture_upload_cost;
+    }
+    if (m_2k_texture_upload_cost.size() == 1024)
+    {
+        S32 index = (S32)llceil(sqrt((F32)area));
+        index -= 1025;
+        index = llclamp(index, 0, 1023);
```

Comment:
It's not looking clean. But I'm not sure how to change it. So let it be.

Reply:
I will add some comments, but hopefully it will change.

Reply:
This fragment was removed